### PR TITLE
[WIP][HttpKernel] Expose bundle hierarchy/metadata as a service

### PR DESCRIPTION
**Pull request by ro0NL**
| Q | A |
| --- | --- |
| Branch? | "master" |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | ~ |
| License | MIT |
| Doc PR | ~ |

This is a proof of concept to expose the bundle hierarchy and metadata in a early phase. For example when needed in a bundle extension. This makes the ecosystem a lot more flexible imho.

``` php
// SomeBundleExtension.php
public function load(array $configs, ContainerBuilder $container) {
    $bundles = $container->get('kernel.bundles');
    $someBundlePath = $bundles['SomeBundle']->getPath();
}
```
- [ ] Useful?
- [ ] Fix caching (service as a definition)
- [ ] Tests
- [ ] Profit.

### CODE REVIEW

**Comment by ro0NL** [comment]
@wesleylancel noted this could perhaps be enhanced using the `bundleMap` property.. however that works a bit counter-wise, as you get the childs per bundle rather then parents, which are needed to construct the root VO properly. I tend to keep it this way..

**Comment by ro0NL** [comment]
Besides that, to fix the caching this parameter needs to be converted some way, some how.

We can go serialization way, or maybe introduce a `ParameterValueInterface` that allows for converting object to array (and vice versa), not sure how we would identify the result in cache =/

File: src/Symfony/Component/HttpKernel/Bundle/BundleVO.php [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+ */
+
+namespace Symfony\Component\HttpKernel\Bundle;
+
+/**
+ * Bundle value object.
+ *
+ * @author Roland Franssen <franssen.roland@gmail.com>
+ */
+final class BundleVO
```
Comment (unkind) [inline src/Symfony/Component/HttpKernel/Bundle/BundleVO.php]: `BundleMetadata`?

Comment (ro0NL) [inline src/Symfony/Component/HttpKernel/Bundle/BundleVO.php]: Yes. Will do.

**Comment by Ener-Getick** [comment]
Why do you need that ?
Isn't `Kernel::locateResource()` enough ?

**Comment by ro0NL** [comment]
Main problem is the `kernel` as a service is not yet available when building the container ([ref](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpKernel/Kernel.php#L494)).

So this approach is about exposing bundle info as a parameter, so it's available in a early phase. Ie when building the container, ie extensions. Besides i like the way this represents your bundle ecosystem, imho it's more useful then _just_ `kernel.bundles`, or `locateResource`.

**Comment by Ener-Getick** [comment]
@ro0NL as you said i think the real problem here is that the kernel is not available during compilation, maybe we should try to do that at first ?

**Comment by ro0NL** [comment]
Good point. However, im not sure we should pass the kernel around, as is, in this phase. I can imagine reasons why it's currently added afterwards, ie you can really break the ecosystem as lots of stuff is publicly accessible, not frozen, etc. Hence i chose for a VO object, to avoid passing bundle instances around at all.

But maybe someone can advice why to, or why not.

**Comment by ro0NL** [comment]
I like the idea of separating the kernel from metadata/utilities, and break it down to just booting/handling/terminating etc. For metadata/utility the real kernel just uses the "kernel as a service".

Eventually im not to sure about passing the real kernel as a service (like now after building), i think it's made available for the wrong reasons (ie exposing metadata/utilities).

**Comment by Ener-Getick** [comment]
@ro0NL many things can break an app, i don't think that because someone can misuse a feature we should not include it.

**Comment by ro0NL** [comment]
Im pretty sure this is the reason it's registered after building. If not, this is as easy as moving a few lines around, so im really curious here...

**Comment by Ener-Getick** [comment]
@ro0NL no that's not that easy because the `ContainerBuilder` removes definitions when you use `set`, so the `kernel` service would never exist.

**Comment by ro0NL** [comment]
I havent fully investigated this yet, im still a bit unaware of inner workings here. Could you perhaps explain "the ContainerBuilder removes definitions when you use set" / point me to it.

Eventually i would like the kernel metadata to be cached as a service definition.

**Comment by Ener-Getick** [comment]
That's very complicated to access services during the compilation, mostly due to [this](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/DependencyInjection/ContainerBuilder.php#L372-L376).
You can't have a definition of a service and the service instance manually defined in the builder.

**Comment by ro0NL** [comment]
Got it. This is hard.. what about an object being a frozen definition and a real service at the same time...

**Comment by Ener-Getick** [comment]
@ro0NL it could be a solution but remember that we have to deal with bc :)

**Comment by ro0NL** [comment]
Of course, i'll have look at it this weekend to work on a better poc.

**Comment by ogizanagi** [comment]
Not convinced by the use cases targeted by this feature, but what about something like:

```php
$container->set('booting_kernel', new BootingKernel($this));
```

where `BootingKernel` is a simple kernel decorator, preventing from using non-allowed methods by throwing a proper exception ?

**Comment by ro0NL** [comment]
The usecase is to have a ecosystem that allows for bugs like #6919 and #18563 to be fixed, and not have to workaround each time.

Im aware of the container builder, and how it works. I tend to go with a `ServiceAwareDefinitionInterface::getService` that allows for `get` only if the corresponding definition implements that... it's an idea ;-)

File: src/Symfony/Component/HttpKernel/Bundle/BundleMetadata.php [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+    /**
+     * Get string representation.
+     *
+     * @return string
+     */
+    public function __toString()
+    {
+        return $this->className;
+    }
```
Comment (iltar) [inline src/Symfony/Component/HttpKernel/Bundle/BundleMetadata.php]: Is a `__toString` really necessary? If you want the class name, you can just call it from the getter

**Comment by iltar** [comment]
I'm not entirely sure what this does yet... But will this allow the addition of a "virtual" bundle? Basically hinting towards 2 things:
 - AppKernel/Bundle merge which removes the need of the AppBundle concept
 - Autogenerating bundle information to map specific vendor packages as bundle without having to do so to prevent recurring code duplication

File: src/Symfony/Component/HttpKernel/Bundle/BundleMetadata.php (hunk 2 above)
Comment (ro0NL) [inline src/Symfony/Component/HttpKernel/Bundle/BundleMetadata.php]: No.. i forgot to remove it. So ignore for now.

**Comment by ro0NL** [comment]
Yes, it should allow for virtual bundles. I dont get your 2 points fully yet, but i try to show a better POC this afternoon.

**Comment by iltar** [comment]
I've taken some snapshots of how we've currently implemented it:
```php
class EntityPackage
{
    public function __construct($name, $path, $namespace = '', $alias = '');
    public function getAlias();
    public function getName();
    public function getPath();
    public function getNamespace();
}

// ...
public static function load(&$bundles, $file = EntityInformation::CACHE_FILE)
{
    $packages = (new EntityInformation($file))->getEntityPackages();

    // Create and append HostnetEntityBundles for all entries.
    foreach ($packages as $package) {
        $bundles[] = new HostnetEntityBundle($package);
    }
    $bundles[] = new HostnetDoctrineBundle();
}

// and in the end of registerBundles() in AppKernel
HostnetEntityBundle::load($bundles);
return $bundles;
```

Basically comes down to having entity packages which feature no bundle concept/dependency while still registering all of them as bundle once in symfony and hooking up their repositories to the DIC by registering them as services.

**Comment by ro0NL** [comment]
Looks creative :-) no offense. Imho this look more event related (listening on kernel boot events that is). My need is to modify the containerbuilder from a bundle extension, based on the ecosystem (kernel info, bundle info, etc.). For exmple the twig [issue](https://github.com/symfony/symfony/pull/19586) which needs to act upon the bundle hierarchy. Having this as a service would be so convenient :-)

**Comment by ogizanagi** [comment]
@ro0NL: Why not simply allow to use `ContainerBuilder::get()` for services set through the `ContainerInterface::set()` method ? 
Those services are safe to use during build time anyway, so the deprecation doesn't make much sense for them, only resolving services definition does, right ?

**Comment by ro0NL** [comment]
Im not sure.. i can imagine this is a design choice. There are 2 phases (building and runtime), where there's only 1 container per phase (respectively `ContainerBuilder` and `Container`) available. Basically in between compiling happens...

However due the 1st phase we cant use it as container throughout as it can guarantee safe services (not fully built) when getting. Why setting direct (fully built) services is not allowed, im not sure.. like i said i guess it's a design choice:

The `ContainerBuilder` can be used _as is_ (the instance you have) as a `Container` only after compiling to ensure safe services, the role of `set()` is that it must matched with an existing defintion. I cant fully explain that, but it makes sense in a way.

Anyway, to adhere to all this, i chose for _true_ service objects representing the ecosystem (ie they have a corresponding definition), which allows it to work in both phases. Eventually this allows also for separation of concerns (the kernel (functional), and kernel-service (informational/util). To clarify, `Kernel::locateResource` is not used by the kernel itself..

**Pull request closed without merging**